File: python/sparktk/graph/ops/label_propagation.py
# vim: set encoding=utf-8

#  Copyright (c) 2016 Intel Corporation 
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
#


def label_propagation(self, max_iterations):
    """

    Parameters
    ----------

    Assigns label based off of proximity to different vertices. The labels
    are initially 1 unique label per vertex (the vertex id), and as the
    algorithm runs some of these get erased

    Note this algorithm is neither guaranteed to converge, nor guaranteed to
    converge to the correct value.

    This calls graph frames label propagation which can be found at 

    http://graphframes.github.io/api/scala/index.html#org.graphframes.lib.LabelPropagation

    :return: (Frame) Frame containing the vertex id's and the community they are a member of

    Examples
    --------

        >>> vertex_schema = [('id', int)]
        >>> vertex_rows = [ [1], [2], [3], [4], [5] ]

        >>> edge_rows = [ [1, 2], [1, 3], [2, 3], [1, 4], [4, 5] ]
        >>> edge_schema = [('src', int), ('dst', int)]

        >>> vertex_frame = tc.frame.create(vertex_rows, vertex_schema)
        >>> edge_frame = tc.frame.create(edge_rows, edge_schema)

        >>> graph = tc.graph.create(vertex_frame, edge_frame)

        >>> result = graph.label_propagation(10)
        >>> result.inspect()
        [#]  id  label
        ==============
        [0]   1      1
        [1]   2      2
        [2]   3      2
        [3]   4      2
        [4]   5      1

    """
    from sparktk.frame.frame import Frame
    return Frame(self._tc, self._scala.labelPropagation(max_iterations))

File: sparktk-core/src/main/scala/org/trustedanalytics/sparktk/graph/internal/ops/LabelPropagation.scala
/**
 *  Copyright (c) 2016 Intel Corporation 
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package org.trustedanalytics.sparktk.graph.internal.ops

import org.trustedanalytics.sparktk.frame.Frame
import org.apache.spark.sql.functions._
import org.graphframes.GraphFrame
import org.graphframes.GraphFrame.ID
import org.apache.spark.sql.DataFrame
import org.graphframes.lib.AggregateMessages

import org.apache.spark.sql.Column
import org.apache.spark.sql.functions.lit

import org.trustedanalytics.sparktk.graph.internal.{ GraphState, GraphSummarization, BaseGraph }

trait LabelPropagationSummarization extends BaseGraph {
  /**
   *
   * Label propagation attempts to determine communities based off of neighbor associations. A community
   * is a label that any vertex can have assigned to it, vertices are assumed to be more likely members
   * of communities they are near.
   *
   * This algorithm can fail to converge, oscillate or return the trivial solution (all members of one
   * community).
   *
   * @param maxIterations the number of iterations to run label propagation for
   * @return dataFrame with the vertices associated with their respective communities
   */
  def labelPropagation(maxIterations: Int): Frame = {
    execute[Frame](LabelPropagation(maxIterations))
  }
}

case class LabelPropagation(maxIterations: Int) extends GraphSummarization[Frame] {
  require(maxIterations > 0, "maxIterations must be a positive value")

  val index = "label"
  val indexNew = "labelNew"

  // This is the column name spark gives to count
  val countName = "count"
  val maxName = "max"
  val maxValue = "max.MSG"

  val messageName = "MSG"

  override def work(state: GraphState): Frame = {

    var graph = GraphFrame(state.graphFrame.vertices.withColumn(index, monotonicallyIncreasingId()), state.graphFrame.edges)

    for (i <- 1 to maxIterations) {
      val updatedComponent =
        GraphHelpers.agg(graph, Some(AggregateMessages.dst(index)), Some(AggregateMessages.src(index)))
          .groupBy(ID, messageName)
          .count()
          .groupBy(ID)
          .agg(max(struct(col(countName),
            col(messageName))).alias(maxName))
          .select(col(ID), col(maxValue))
          .withColumnRenamed(messageName, indexNew)

      val joinedComponent = updatedComponent
        .join(graph.vertices, graph.vertices(ID) === updatedComponent(ID))
        .drop(graph.vertices(ID))

      val newVertices = joinedComponent
        .drop(index)
        .withColumnRenamed(indexNew, index)

      val unCachedVertices = AggregateMessages.getCachedDataFrame(newVertices)

      graph = GraphFrame(unCachedVertices, state.graphFrame.edges)
    }

    new Frame(graph.vertices)
  }
}
